packages/@aws-cdk/aws-kinesisanalyticsv2/lib/flink-application.ts

```diff
@@ -0,0 +1,328 @@
+
+    return new Import(scope, id);
+  }
+
+  public readonly applicationArn: string;
+  public readonly applicationName: string;
+
+  // Role must be optional for JSII compatibility
+  public readonly role?: iam.IRole;
+
+  public readonly grantPrincipal: iam.IPrincipal;
+
+  constructor(scope: Construct, id: string, props: FlinkApplicationProps) {
+    super(scope, id, { physicalName: props.applicationName });
+    validateFlinkApplicationProps(props);
+
+    this.role = props.role ?? new iam.Role(this, 'Role', {
+      assumedBy: new iam.ServicePrincipal('kinesisanalytics.amazonaws.com'),
+    });
+    this.grantPrincipal = this.role;
+
+    // Permit metric publishing to CloudWatch
+    this.role.addToPrincipalPolicy(new iam.PolicyStatement({
+      actions: ['cloudwatch:PutMetricData'],
+      resources: ['*'],
+    }));
+
+    // TODO: Refactor to make this bug impossible
+    const code = props.code.bind(this);
+    if (!props.code.bucket) {
+      throw new Error('BUG: Bound code is missing its bucket.');
+    }
+    props.code.bucket.grantRead(this);
+
+    const resource = new CfnApplication(this, 'Resource', {
+      runtimeEnvironment: props.runtime.value,
+      serviceExecutionRole: this.role.roleArn,
+      applicationConfiguration: {
+        applicationCodeConfiguration: code,
+        environmentProperties: props.propertyGroups?.length
+          ? { propertyGroups: props.propertyGroups.map(pg => pg.toCfn()) }
+          : undefined,
+        flinkApplicationConfiguration: flinkApplicationConfiguration({
+          checkpointingEnabled: props.checkpointingEnabled,
+          checkpointInterval: props.checkpointInterval,
+          minPauseBetweenCheckpoints: props.minPauseBetweenCheckpoints,
+          logLevel: props.logLevel,
+          metricsLevel: props.metricsLevel,
+          autoScalingEnabled: props.autoScalingEnabled,
+          parallelism: props.parallelism,
+          parallelismPerKpu: props.parallelismPerKpu,
+        }),
+        applicationSnapshotConfiguration: {
+          snapshotsEnabled: props.snapshotsEnabled ?? true,
+        },
+      },
+    });
+    resource.node.addDependency(this.role);
+
+    const logGroup = new logs.LogGroup(this, 'LogGroup', {
+      retention: logs.RetentionDays.ONE_WEEK,
+      removalPolicy: core.RemovalPolicy.DESTROY,
+    });
+
+    const logStream = new logs.LogStream(this, 'LogStream', {
+      logGroup,
+      removalPolicy: core.RemovalPolicy.DESTROY,
+    });
+
+    const logStreamArn = `arn:${core.Aws.PARTITION}:logs:${core.Aws.REGION}:${core.Aws.ACCOUNT_ID}:log-group:${logGroup.logGroupName}:log-stream:${logStream.logStreamName}`;
+
+    // Permit logging
+    this.role.addToPrincipalPolicy(new iam.PolicyStatement({
+      actions: ['logs:DescribeLogStreams', 'logs:DescribeLogGroups'],
+      resources: [
+        logGroup.logGroupArn,
+        core.Stack.of(this).formatArn({
+          service: 'logs',
+          resource: 'log-group',
+          resourceName: '*',
```

Comment:
When I first set up a Flink application on AWS I thought I could reason through these permissions better than what was shown in the AWS docs. However I couldn't really figure out why things weren't working so I gave up and copy pasted [the permissions show in the docs](https://docs.aws.amazon.com/kinesisanalytics/latest/java/cloudwatch-logs.html#enable_putlogevents_permissions_policy).

The author of the [example I linked](https://github.com/aws-samples/amazon-kinesis-analytics-streaming-etl/blob/master/cdk/lib/streaming-etl.ts#L100) seems to be more IAM savvy than me and the docs author so I copied their approach. The resource ARNs end up looking like:

```
// I think this one is used for logs:DescribeLogStreams
// logGroup.logGroupArn is an oddball case with a :* on the end
arn:aws:logs:us-east-1:123456789012:log-group:my-log-group:*
// I think this one applies to logs:DescribeLogGroups
arn:aws:logs:us-east-1:123456789012:log-group:*
```

It might be possible to split these two actions apart, it would just take a bit of time to manually test. Do you have a preference for the official documentation, the AWS sample app, or trying to break this apart further?

Reply:
I believe the difference is:

1. `arn:aws:logs:us-east-1:123456789012:log-group:my-log-group:*` refers to the Log Streams inside the Log Group with the name `my-log-group`.
2. `arn:aws:logs:us-east-1:123456789012:log-group:*` refers to all Log Groups within the account and region.

So I think splitting these up is probably a good idea.

Reply:
I made this change and was able to verify that the logging continued to work.